Tests/NIOTests/CodecTest.swift

```diff
@@ -1561,30 +1588,361 @@ public final class MessageToByteEncoderTest: XCTestCase {
+                self.decodeLastCalls += 1
+                XCTAssertEqual(1, self.decodeLastCalls)
+                XCTAssertEqual(0, buffer.readableBytes)
+                return ()
+            }
+        }
+
+        let decoder = Decoder()
+        let processor = NIOSingleStepByteToMessageProcessor(decoder)
+        let messageReceiver: MessageReceiver<()> = MessageReceiver()
+
+        XCTAssertEqual(0, messageReceiver.count)
+
+        XCTAssertNoThrow(try processor.finishProcessing(seenEOF: true, messageReceiver.receiveMessage))
+        XCTAssertNotNil(messageReceiver.retrieveMessage())
+        XCTAssertNil(messageReceiver.retrieveMessage())
+
+        XCTAssertEqual(1, decoder.decodeLastCalls)
+    }
+
+    func testPayloadTooLarge() {
+        struct Decoder: NIOSingleStepByteToMessageDecoder {
+            typealias InboundOut = Never
+
+            func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
+                return nil
+            }
+
+            func decodeLast(buffer: inout ByteBuffer, seenEOF: Bool) throws -> InboundOut? {
+                return nil
+            }
+        }
+
+        let max = 100
+        let allocator = ByteBufferAllocator()
+        let processor = NIOSingleStepByteToMessageProcessor(Decoder(), maximumBufferSize: max)
+        let messageReceiver: MessageReceiver<Never> = MessageReceiver()
+
+        var buffer = allocator.buffer(capacity: max + 1)
+        buffer.writeString(String(repeating: "*", count: max + 1))
+        XCTAssertThrowsError(try processor.process(buffer: buffer, messageReceiver.receiveMessage)) { error in
+            XCTAssertTrue(error is ByteToMessageDecoderError.PayloadTooLargeError)
+        }
+    }
+
+    func testPayloadTooLargeButHandlerOk() {
+        class Decoder: NIOSingleStepByteToMessageDecoder {
+            typealias InboundOut = String
+
+            var decodeCalls = 0
+
+            func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
+                self.decodeCalls += 1
+                return buffer.readString(length: buffer.readableBytes)
+            }
+
+            func decodeLast(buffer: inout ByteBuffer, seenEOF: Bool) throws -> InboundOut? {
+                return try decode(buffer: &buffer)
+            }
+        }
+
+        let max = 100
+        let allocator = ByteBufferAllocator()
+        let decoder = Decoder()
+        let processor = NIOSingleStepByteToMessageProcessor(decoder, maximumBufferSize: max)
+        let messageReceiver: MessageReceiver<String> = MessageReceiver()
+
+        var buffer = allocator.buffer(capacity: max + 1)
+        buffer.writeString(String(repeating: "*", count: max + 1))
+        XCTAssertNoThrow(try processor.process(buffer: buffer, messageReceiver.receiveMessage))
+        XCTAssertNoThrow(try processor.finishProcessing(seenEOF: false, messageReceiver.receiveMessage))
+        XCTAssertEqual(0, processor._buffer!.readableBytes)
+        XCTAssertGreaterThan(decoder.decodeCalls, 0)
```

Comment:
could we add one test that does cause re-entrancy? I know it's not an issue but we should prove it.